pdf/pdfselection.cpp

```diff
@@ -267,51 +267,78 @@ void PDFSelection::Private::textBoxAtPoint(const QPointF &point, Position positi
+        PDFSelection::Private::Position currentBoxPosition;
+        QRectF previousBox {0, -1, 1, 1};
+        qreal closestSquaredDistance = 2.;
+        *boxIndex = boxes.length();
         for (PDFDocument::TextList::const_iterator box = boxes.begin();
              box != boxes.end(); box++) {
-            /* Stop counting boxes down as soon as a box after reducedCoordPoint is found. */
-            if (box->first.y() > reducedCoordPoint.y() ||
-                (box->first.y() + box->first.height() > reducedCoordPoint.y() &&
-                 box->first.x() + box->first.width() > reducedCoordPoint.x())) {
-                return;
+            currentBoxPosition = (box->first.bottom() < reducedCoordPoint.y() ||
+                                  (box->first.top() < reducedCoordPoint.y() &&
+                                   box->first.left() < reducedCoordPoint.x()))
+                ? PDFSelection::Private::Before
+                : PDFSelection::Private::After;
+            // Find a transition in the list of boxes. A transition is
+            // when the current box switch from being before the given point
+            // to being after.
+            if (previousBoxPosition == PDFSelection::Private::Before &&
+                currentBoxPosition == PDFSelection::Private::After) {
+                // The actual transition that will be kept is the closest
+                // to the given point.
+                QPointF boxCenter = previousBox.center();
+                qreal squaredDistance =
+                    (boxCenter.x() - reducedCoordPoint.x()) *
+                    (boxCenter.x() - reducedCoordPoint.x()) +
+                    (boxCenter.y() - reducedCoordPoint.y()) *
+                    (boxCenter.y() - reducedCoordPoint.y());
+                if (squaredDistance < closestSquaredDistance) {
+                    *boxIndex = index;
+                    closestSquaredDistance = squaredDistance;
+                }
+                boxCenter = box->first.center();
+                squaredDistance =
+                    (boxCenter.x() - reducedCoordPoint.x()) *
+                    (boxCenter.x() - reducedCoordPoint.x()) +
+                    (boxCenter.y() - reducedCoordPoint.y()) *
+                    (boxCenter.y() - reducedCoordPoint.y());
+                if (squaredDistance < closestSquaredDistance) {
+                    *boxIndex = index;
+                    closestSquaredDistance = squaredDistance;
+                }
             }
-            *boxIndex += 1;
+            previousBoxPosition = currentBoxPosition;
+            previousBox = box->first;
+            index += 1;
+        }
+        if (position == PDFSelection::Private::Before ||
+            (position == PDFSelection::Private::After &&
+             *boxIndex > 0 && *boxIndex <= boxes.length() &&
+             boxes[*boxIndex - 1].first.contains(reducedCoordPoint)))
+            *boxIndex -= 1;
```

Comment:
Usually not that important, but here operators at line start and braces with multi-line if statement could help readability.